core/src/HIP/Kokkos_HIP_Instance.cpp

```diff
@@ -372,3 +372,13 @@ void hip_internal_error_throw(hipError_t e, const char *name, const char *file,
 }
 }  // namespace Impl
 }  // namespace Kokkos
+
+//----------------------------------------------------------------------------
+
+namespace Kokkos {
+namespace Experimental {
+HIP::size_type HIP::detect_device_count() {
+  return HIPInternalDevices::singleton().m_hipDevCount;
+}
```

Comment:
Why did you implement it here rather than `Kokkos_HIP_Space.cpp`?

Reply:
`HIPInternalDevices` is declared in this file only in this file in an anonymous namespace. If preferred, we can also declare that class in `Kokkos_HIP_Instance.hpp` and move the implementation to `Kokkos_HIP_Space.cpp`.

Reply:
Fair enough.  We will need to clean that out at some point.  I am not asking you to change.
https://github.com/kokkos/kokkos/blob/266724dd8405bc41e056ac2c473b23de1d977554/core/src/HIP/Kokkos_HIP_Space.cpp#L590-L593
https://github.com/kokkos/kokkos/blob/266724dd8405bc41e056ac2c473b23de1d977554/core/src/HIP/Kokkos_HIP_Space.cpp#L631-L633